### PR TITLE
Skim improvements

### LINKED ISSUE #3438: Skims improvements: Transit, ridehail, parking
We need a few improvements to transit skims for the activitysim/PILATES integration:

## Transit

- [x] **Wait times**: Right now there is no wait time included in the skims for the first transit leg because we delay the departure of the first walk leg so that it arrives at the transit station at the exact time the transit vehicle leaves. This means we don't capture any delay between when the trip _wants_ to leave and when it actually does. Can we change the timing of the initial walk access leg so that it leaves at the previous activityEndTime, so there's a gap between the walk leg ending and the next transit leg starting?
- [x] **Denied boarding**: If a person tries to take a transit trip but then cannot complete it because a vehicle is full (and they then need to replan), we should include a way of having the skims reflect that the original trip wasn't possible. Two ideas are 1) if they do replan and take transit again (e.g. if they wait for the next bus), we should make sure that the travel time and wait time from the original trip get accounted for in the skim as well, 2) We come up with a "penalty" skim value that we record for the initial OD that increases the expected travel time for that origin/destination/mode/time bin combo.
- [x] **No route available**: Similar to above, if a person has in their plans that they should take transit for a trip, but the router does not return any feasible transit routes, we should record a penalty value to the OD skimmer with a high travel time and/or cost so that ActivitySim knows not to send any people on transit for that origin/destination

## 1-d skims for ridehail and parking
There's some data needed by activitysim that doesn't need to be at the OD level but rather only needs to have one record for every TAZ and time period
- [x] **Ride-hail stats**: For every TAZ and time period (hour?), we want to record the average wait time, average cost per mile, and % of requests that go unmatched for ridehail requests originating from that TAZ, for solo and pooled ridehail separately
- [x] **Parking and charging stats**: For every TAZ and time period, we should also record the average walk access distance and parking cost per hour. Note that we probably want this to be recorded for the TAZ where the next activity is located (not necessarily where the parking spot ends up), and we want to break it down by charger type at the parking spot (e.g. maybe having three categories: No Charger, AC Slow Charger, DC Fast Charger)

## Freight (2-d skims)
- [x] **travel Time**, **distance**, **energy** etc: The same way we are calculating them for other travel modes. 
- [x] **payload weight in Kg**: The average amount of goods in Kg being transported from origin to destination.
## Freight (1-d skim)
- [x] **Freight stat**: for every TAZ and time period, sum freight stops (loading and unloading), average cost per mile, average walk access distance and parking cost per hour. (same as above, for activity and not parking spot)

### PR DESCRIPTION
**Pull request by dimaopen**
Added failedTrips field to ODSkims. One can use failedTrips/observations ratio to evaluate probabilty of succesful trip.
If trip is TRANSIT then the trip starts right at the activity ends in order to take into account waiting for the transit vehicle.

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3447)

### CODE REVIEW

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
Comment (zneedell) [inline src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala]: Just out of curiosity--would we expect this to give different behavior if instead of the tick we gave the previous activity end time? Would that be more robust to parallelism issues and/or introduce other problems?

**Review by zneedell — COMMENTED** [review]
One non-blocking question, but LGTM other than that.

I think it probably makes sense to do what you have here and not directly penalize failed trips in the skims (but rather to just keep track of failed trips and penalize on ActivitySim's side), but I'll confirm with them that that should work

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
Comment (dimaopen) [reply src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala]: This tick is the same as the Activity end time. But we could check that the Activity end time is not in the past and just use it here. I'll change it. I can imagine that the Mode Choice may happen not at the activity end time but at some time before that. In this case, using activity end time is better. But we shouldn't use a tick in the past here because it causes issues with the scheduler.

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Review by zneedell — APPROVED** [review]
LGTM pending any comments from @JustinPihony

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala]: To help with readability and understanding why this is considered a failed event I suggest rename `mode` to `expectedMode` - or maybe turn the `exists` into a `def` to control the readability?

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:1528]: If I am reading this correct, then any trip that is not a teleportation event should be considered a failure?

File: src/main/scala/beam/router/skim/CsvSkimReader.scala
Comment (JustinPihony) [inline src/main/scala/beam/router/skim/CsvSkimReader.scala]: Doesn't the second parameter do this already?

**Review by JustinPihony — COMMENTED** [review]
There is one question I'd like cleared up about failedTrips before approving

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
Comment (dimaopen) [reply src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:1528]: Generally, this is true. If we have some trip defined (`currentTrip` is an `Option`) that we cannot finish (we're choosing a new trip now (in `FinishingModeChoice` state)) then the currentTrip is considered a failure.
About teleportation, I'm not sure but it seems that they suppose that no teleportation failure is possible.

**Comment by dimaopen** [comment]
Test!

File: src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala
Comment (JustinPihony) [reply src/main/scala/beam/agentsim/agents/modalbehaviors/ChoosesMode.scala:1528]: I will defer to @zneedell on that aspect - and he already approved - so otherwise LGTM

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test!

**Comment by dimaopen** [comment]
Test now!

**Pull request merged**
